### PR TITLE
Send to route rpc #188

### PR DESCRIPTION
**Pull request by t4sk**
* This commit adds a RPC command to send a payment through a user defined route
* Addresses #188
* Related work:  #219, @cfromknecht's branch [swapz](https://github.com/cfromknecht/lnd/tree/swapz)

At the moment `SendToRoute` DOES NOT WORK with the output of `QueryRoutes` for 2 reasons:
* `QueryRoutes` returns hops with time lock delta of 9. But [htlcswitch/link.go](https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/link.go#L1608) checks for a time lock delta of 144.
* `QueryRoutes` returns a fee in satoshi. [hltcswitch/link.go](https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/link.go#L1737) checks fee in millisatoshi. When I try to send a 2-hop payment of 1000 satoshi, a fee of 1001 millisatoshi is expected but the output from `QueryRoute` converts the fee to 1000 millisatoshi, which makes the fee insufficient.

Tests in `lnd_test` are written with the issues above in mind. But it would be ideal if the output of `QueryRoutes` can be used as an input for `SendToRoute`.

This is my first PR, feedback welcomed. 

Thanks to @afederigo and @cfromknecht for the heavy lifting :)

### CODE REVIEW

**Comment by Roasbeef** [comment]
> QueryRoutes returns hops with time lock delta of 9. But htlcswitch/link.go checks for a time lock delta of 144.

A timelock delta of 9 is the default value for the _final_ hop assuming that they don't specify anything in the invoice. Our default policy is to use a value of 144, hence the check in link.go (which isn't hard coded but part of the current policy and can be changed on a per-channel basis). So seems like the solution here is to take a `final_cltv_value` along with `queryroutes`. 

> QueryRoutes returns a fee in satoshi. hltcswitch/link.go checks fee in millisatoshi. When I try to send a 2-hop payment of 1000 satoshi, a fee of 1001 millisatoshi is expected but the output from QueryRoute converts the fee to 1000 millisatoshi, which makes the fee insufficient.

Yep, so `queryroutes` should be modified to return everything in `mSAT`.

**Comment by t4sk** [comment]
@Roasbeef  Thanks for your feedback! It's very informative 👍
 
But can you clarify what needs to be done here.

> So seems like the solution here is to take a final_cltv_value along with queryroutes.

Should `queryroutes` be modified to take a `final_cltv_delta` ?

**Comment by Roasbeef** [comment]
> Should queryroutes be modified to take a final_cltv_delta ?

Yeah, so it would be an _optional_ argument. There's an "assumed" default of 9 blocks within the network. In scenarios where the use case is to pipe the result through `sendtoroute`, then the user would know what the `final_cltv_delta` value should be.

File: routing/router.go [hunk 1]
```diff
@@ -1723,6 +1723,186 @@ func (r *ChannelRouter) SendPayment(payment *LightningPayment) ([32]byte, *Route
...
 }
 
+// SendToRoute attempts to send a payment as described within the passed
+// LightningPayment through the provided routes. This function is blocking
+// and will return either: when the payment is successful, or all routes
+// have been attempted and resulted in a failed payment. If the payment
+// succeeds, then a non-nil Route will be returned which describes the
+// path the successful payment traversed within the network to reach the
+// destination. Additionally, the payment preimage will also be returned.
+func (r *ChannelRouter) SendToRoute(routes []*Route,
```
Comment (halseth) [inline routing/router.go]: This adds a lot of code duplication. Instead should `SendPayment` be modified to use `SendToRoute` internally.

File: rpcserver.go [hunk 2]
```diff
@@ -1981,6 +1990,250 @@ func (r *rpcServer) SendPaymentSync(ctx context.Context,
 	}, nil
 }
 
+// SendToRoute dispatches a bi-directional streaming RPC for sending payments
+// through the Lightning Network via predefined routes passed in. A single RPC
+// invocation creates a persistent bi-directional stream allowing clients to
+// rapidly send payments through the Lightning Network with a single
+// persistent connection.
+func (r *rpcServer) SendToRoute(sendToStream lnrpc.Lightning_SendToRouteServer) error {
```
Comment (halseth) [inline rpcserver.go]: here we should also try to reuse common code from `SendPayment`, an approach is mentioned in the comment here: https://github.com/lightningnetwork/lnd/pull/219/files#diff-01ada5680aaa18fc79cbf78b2d55a37bR1683

File: routing/router.go (hunk 1 above)
Comment (t4sk) [reply routing/router.go]: @halseth thanks for reviewing! I've considered your suggestion but decided to create a private method `sendPayment` which is used by `SendPayment` and `SendToRoute`.

**Comment by t4sk** [comment]
Feedback addressed. Please re-review :)

File: rpcserver.go [hunk 3]
```diff
@@ -2813,16 +2813,16 @@ func (r *rpcServer) QueryRoutes(ctx context.Context,
 func marshallRoute(route *routing.Route) *lnrpc.Route {
 	resp := &lnrpc.Route{
 		TotalTimeLock: route.TotalTimeLock,
-		TotalFees:     int64(route.TotalFees.ToSatoshis()),
-		TotalAmt:      int64(route.TotalAmount.ToSatoshis()),
+		TotalFees:     int64(route.TotalFees),
```
Comment (halseth) [inline rpcserver.go]: The changes to `QueryRoutes` could be made its own PR, with its own set of added tests. Will make it easier to review! 😀

Comment (t4sk) [reply rpcserver.go]: @halseth Should `lnrpc.Route` be extended with additional fields `TotalFeesMSat` and `TotalAmtMSat`? Thanks!

Comment (halseth) [reply rpcserver.go]: Hm, good question. Adding the fields wouldn't break usage of the protos, while changing the names would. So maybe just add them? :)

**Comment by Roasbeef** [comment]
#1081 has been merged now. This needs a rebase now.

**Comment by t4sk** [comment]
@Roasbeef Before rebasing this PR,  should I create a separate PR for the commits which modifies `queryroutes` to take `final_cltv_delta` as an optional param?

**Comment by Roasbeef** [comment]
@t4sk nah I think the commit structure is fine as is.

**Comment by Roasbeef** [comment]
Friendly ping on rebase 🤗

**Comment by t4sk** [comment]
@Roasbeef Rebasing `rpcserver.go` has been giving me a headache. Do you mind if I delete some commits and rework them?

File: rpcserver.go [hunk 4]
```diff
@@ -1607,20 +1616,264 @@ func validatePayReqExpiry(payReq *zpay32.Invoice) error {
...
+	send func(*lnrpc.SendResponse) error
 }
 
 // SendPayment dispatches a bi-directional streaming RPC for sending payments
 // through the Lightning Network. A single RPC invocation creates a persistent
 // bi-directional stream allowing clients to rapidly send payments through the
 // Lightning Network with a single persistent connection.
-func (r *rpcServer) SendPayment(paymentStream lnrpc.Lightning_SendPaymentServer) error {
+func (r *rpcServer) SendPayment(stream lnrpc.Lightning_SendPaymentServer) error {
+	return r.sendPayment(&paymentStream{
```
[the remark this answers is not in the document]
Comment (t4sk) [reply rpcserver.go]: I've tried your suggestion but ran into one problem:
- cannot pass a common interface for `stream` to  `sendPayment` because `lnrpc.Lightning_SendPaymentServer` and `lnrpc.Lightning_SendToRouteServer` are also an interface.

File: routing/router.go [hunk 5]
```diff
@@ -1458,6 +1453,66 @@ func (r *ChannelRouter) SendPayment(payment *LightningPayment) ([32]byte, *Route
...
+		newLogClosure(func() string {
+			payment.Target.Curve = nil
+			return spew.Sdump(payment)
+		}),
+	)
+
+	// We'll continue until our payment succeeds, we encounter a critical error
+	// or all routes failed to send the payment.
+	i := -1
+	getRoute := func(paySession *paymentSession) (*Route, error) {
```
[the remark this answers is not in the document]
Comment (t4sk) [reply routing/router.go]: Great idea! Done that. Please review.

**Comment by Roasbeef** [comment]
Thanks so much for getting this started! I've added some minor style related commits on top, and committed it as `e4e8b86137a43b21ac169f0f13de3a504ab87b30`.
